File: examples/lights/sample_phong_lighting.cpp

#include <CEngine.h>

int main()
{
    auto _app = new engine::COpenGLApp();
    _app->init();

    auto _cameraProjData = engine::CCameraProjData();
    _cameraProjData.projection  = engine::eCameraProjection::PERSPECTIVE;
    _cameraProjData.fov         = 45.0f;
    _cameraProjData.aspect      = engine::COpenGLApp::GetWindow()->aspect();
    _cameraProjData.zNear       = 0.1f;
    _cameraProjData.zFar        = 100.0f;

    auto _camera = new engine::COrbitCamera( "orbit",
                                             { 0.0f, 0.0f, 3.0f },
                                             { 0.0f, 0.0f, 0.0f },
                                             engine::eAxis::Z,
                                             _cameraProjData,
                                             engine::COpenGLApp::GetWindow()->width(),
                                             engine::COpenGLApp::GetWindow()->height() );

//     const float _cameraSensitivity  = 0.25f;
//     const float _cameraSpeed        = 25.0f;
//     const float _cameraMaxDelta     = 10.0f;
//     auto _camera = new engine::CFpsCamera( "fps",
//                                            { 0.0f, 0.0f, 3.0f },
//                                            { 0.0f, 0.0f, 0.0f },
//                                            engine::eAxis::Z,
//                                            _cameraProjData,
//                                            _cameraSensitivity,
//                                            _cameraSpeed,
//                                            _cameraMaxDelta );

    auto _box = engine::CMeshBuilder::createBox( 3.0f, 3.0f, 3.0f );
    auto _sphere = engine::CMeshBuilder::createSphere( 1.5f );
    auto _gizmo = engine::CMeshBuilder::createBox( 0.2f, 0.2f, 0.2f );
    _gizmo->position = { 0.0f, 0.0f, 2.0f };

    /* load the shader used for this example */
    std::string _baseNamePhong = std::string( ENGINE_EXAMPLES_PATH ) + "lights/shaders/phong";
    auto _shaderPhong = engine::CShaderManager::CreateShaderFromFiles( "phong_shader",
                                                                       _baseNamePhong + "_vs.glsl",
                                                                       _baseNamePhong + "_fs.glsl" );

    std::string _baseNameGouraud = std::string( ENGINE_EXAMPLES_PATH ) + "lights/shaders/gouraud";
    auto _shaderGouraud = engine::CShaderManager::CreateShaderFromFiles( "gouraud_shader",
                                                                         _baseNameGouraud + "_vs.glsl",
                                                                         _baseNameGouraud + "_fs.glsl" );

    ENGINE_ASSERT( _shaderPhong, "Could not load phong shader for our tests :(" );
    ENGINE_ASSERT( _shaderGouraud, "Could not load gouraud shader for our tests :(" );

    /* grab a simple shader to render the camera gizmo */
    auto _shaderGizmo = engine::CShaderManager::GetCachedShader( "basic3d_no_textures" );
    ENGINE_ASSERT( _shaderGizmo, "Could not grab the basic3d shader to render the light gizmo :(" );

    // select shader to use
    auto _shaderLighting = _shaderPhong;
    auto _mesh = _sphere;
    // _mesh->position = { 1.0f, 1.0f, 0.0f };
    // _mesh->scale = { 0.5f, 1.0f, 1.5f };

    bool _moveLight = false;
    float _mvParam = 0.0f;

    while( _app->active() )
    {
        if ( engine::CInputHandler::IsKeyDown( ENGINE_KEY_ESCAPE ) )
        {
            break;
        }
        else if ( engine::CInputHandler::CheckSingleKeyPress( ENGINE_KEY_S ) )
        {
            _shaderLighting = ( _shaderLighting->name() == "phong_shader" ) ? _shaderGouraud : _shaderPhong;
            ENGINE_INFO( "Using shader: {0}", _shaderLighting->name() );
        }
        else if ( engine::CInputHandler::CheckSingleKeyPress( ENGINE_KEY_P ) )
        {
            _moveLight = !_moveLight;
            ENGINE_INFO( "Light state: {0}", ( _moveLight ) ? "moving" : "fixed" );
        }

        engine::CDebugDrawer::DrawLine( { 0.0f, 0.0f, 0.0f }, { 5.0f, 0.0f, 0.0f }, { 1.0f, 0.0f, 0.0f } );
        engine::CDebugDrawer::DrawLine( { 0.0f, 0.0f, 0.0f }, { 0.0f, 5.0f, 0.0f }, { 0.0f, 1.0f, 0.0f } );
        engine::CDebugDrawer::DrawLine( { 0.0f, 0.0f, 0.0f }, { 0.0f, 0.0f, 5.0f }, { 0.0f, 0.0f, 1.0f } );

        _app->begin();
        _camera->update();

        if ( _moveLight )
        {
            _mvParam += 100.0f * engine::CTime::GetTimeStep();
            // _gizmo->position.x = 1.0f + std::sin( _mvParam ) * 2.0f;
            // _gizmo->position.y = std::sin( _mvParam / 2.0f ) * 1.0f;

            _gizmo->position.x = 10.0f * std::sin( _mvParam );
            _gizmo->position.y = 10.0f * std::cos( _mvParam );
            _gizmo->position.z = 0.0f;
        }

        /* do our thing here ************************/
        _shaderLighting->bind();
        _shaderLighting->setMat4( "u_modelMat", _mesh->matModel() );
        _shaderLighting->setMat4( "u_viewProjMat", _camera->matProj() * _camera->matView() );
        _shaderLighting->setMat4( "u_normalMat", ( ( _mesh->matModel() ).inverse() ).transpose() );
        _shaderLighting->setVec3( "u_objectColor", { 1.0f, 0.5f, 0.31f } );
        _shaderLighting->setVec3( "u_lightColor", { 1.0f, 1.0f, 1.0f } );
        _shaderLighting->setVec3( "u_lightPosition", _gizmo->position );
        _shaderLighting->setVec3( "u_viewerPosition", _camera->position() );

        _mesh->render();

        _shaderLighting->unbind();

        _shaderGizmo->bind();
        _shaderGizmo->setMat4( "u_tModel", _gizmo->matModel() );
        _shaderGizmo->setMat4( "u_tView", _camera->matView() );
        _shaderGizmo->setMat4( "u_tProj", _camera->matProj() );
        _shaderGizmo->setVec3( "u_color", { 1.0f, 1.0f, 1.0f } );

        _gizmo->render();

        _shaderGizmo->unbind();
        /********************************************/

        // engine::CDebugDrawer::DrawNormals( _mesh, { 0.0f, 0.0f, 1.0f } );
        engine::CDebugDrawer::Render( _camera );

        _app->end();

        // ENGINE_TRACE( "frame-time: {0}", engine::CTime::GetRawTimeStep() );
    }

    return 0;
}

File: examples/python/lights/sample_phong_lighting.py
#!/usr/bin/env python

import tinyrenderer as tr
import tinymath as tm
import numpy as np
from OpenGL.GL import *

if __name__ == '__main__' :
    app = tr.Application()

    cameraProjData = tr.CameraProjData()
    cameraProjData.projection  = tr.CameraProjection.PERSPECTIVE
    cameraProjData.fov         = 45.0
    cameraProjData.aspect      = app.window().aspect
    cameraProjData.zNear       = 0.1
    cameraProjData.zFar        = 100.0

    camera = tr.OrbitCamera( 'orbit',
                             [ 0.0, 0.0, 3.0 ],
                             [ 0.0, 0.0, 0.0 ],
                             tr.Axis.Z,
                             cameraProjData,
                             app.window().width,
                             app.window().height )

    #### cameraSensitivity  = 0.25
    #### cameraSpeed        = 25.0
    #### cameraMaxDelta     = 10.0
    #### camera = tr.FpsCamera( 'fps',
    ####                        [ 0.0, 0.0, 3.0 ],
    ####                        [ 0.0, 0.0, 0.0 ],
    ####                        tr.Axis.Z,
    ####                        cameraProjData,
    ####                        cameraSensitivity,
    ####                        cameraSpeed,
    ####                        cameraMaxDelta )

    box = tr.MeshBuilder.CreateBox( 3.0, 3.0, 3.0 )
    sphere = tr.MeshBuilder.CreateSphere( 1.5 )
    gizmo = tr.MeshBuilder.CreateBox( 0.2, 0.2, 0.2 )
    gizmo.position = [ 0.0, 0.0, 2.0 ]

    # load the shader used for this example
    baseNamePhong = tr.ENGINE_EXAMPLES_PATH + 'lights/shaders/phong'
    shaderPhong = tr.ShaderManager.CreateShaderFromFiles( 'phong_shader',
                                                          baseNamePhong + '_vs.glsl',
                                                          baseNamePhong + '_fs.glsl' )

    baseNameGouraud = tr.ENGINE_EXAMPLES_PATH + 'lights/shaders/gouraud'
    shaderGouraud = tr.ShaderManager.CreateShaderFromFiles( 'gouraud_shader',
                                                            baseNameGouraud + '_vs.glsl',
                                                            baseNameGouraud + '_fs.glsl' )

    assert shaderPhong, 'Could not load phong shader for our tests :('
    assert shaderGouraud, 'Could not load gouraud shader for our tests :('

    # grab a simple shader to render the camera gizmo
    shaderGizmo = tr.ShaderManager.GetCachedShader( "basic3d_no_textures" )
    assert shaderGizmo, 'Could not grab the basic3d shader to render the light gizmo :('

    # select shader to use
    shaderLighting = shaderPhong
    mesh = sphere
    # mesh.position = [ 1.0, 1.0, 0.0 ]
    # mesh.scale = [ 0.5, 1.0, 1.5 ]

    moveLight = False
    mvParam = 0.0

    while( app.active() ) :
        if ( tr.InputManager.IsKeyDown( tr.Keys.KEY_ESCAPE ) ) :
            break
        elif ( tr.InputManager.CheckSingleKeyPress( tr.Keys.KEY_S ) ) :
            shaderLighting = shaderGouraud if ( shaderLighting.name == 'phong_shader' ) else shaderPhong
            print( 'Using shader: {}'.format( shaderLighting.name ) )
        elif ( tr.InputManager.CheckSingleKeyPress( tr.Keys.KEY_P ) ) :
            moveLight = not moveLight
            print( 'Light state: {}'.format( 'moving' if moveLight else 'fixed' ) )

        tr.DebugDrawer.DrawLine( [ 0.0, 0.0, 0.0 ], [ 5.0, 0.0, 0.0 ], [ 1.0, 0.0, 0.0 ] )
        tr.DebugDrawer.DrawLine( [ 0.0, 0.0, 0.0 ], [ 0.0, 5.0, 0.0 ], [ 0.0, 1.0, 0.0 ] )
        tr.DebugDrawer.DrawLine( [ 0.0, 0.0, 0.0 ], [ 0.0, 0.0, 5.0 ], [ 0.0, 0.0, 1.0 ] )

        app.begin()
        camera.update()

        if ( moveLight ) :
            mvParam += 10.0 * tr.Time.GetTimeStep()
            #### gizmo.position = [ 1.0 + np.sin( mvParam ) * 2.0, np.sin( mvParam / 2.0 ) * 1.0 ]
            gizmo.position = [ 10.0 * np.sin( mvParam ), 10.0 * np.cos( mvParam ), 0.0 ]

        # do our thing here ######################################################
        shaderLighting.bind()
        shaderLighting.setMat4( "u_modelMat", mesh.matModel() )
        shaderLighting.setMat4( "u_viewProjMat", np.dot(camera.matProj(), camera.matView() ) )
        shaderLighting.setMat4( "u_normalMat", tm.inverse( mesh.matModel() ).transpose() )
        shaderLighting.setVec3( "u_objectColor", [ 1.0, 0.5, 0.31 ] )
        shaderLighting.setVec3( "u_lightColor", [ 1.0, 1.0, 1.0 ] )
        shaderLighting.setVec3( "u_lightPosition", gizmo.position )
        shaderLighting.setVec3( "u_viewerPosition", camera.position )

        mesh.render()

        shaderLighting.unbind()

        shaderGizmo.bind()
        shaderGizmo.setMat4( 'u_tModel', gizmo.matModel() )
        shaderGizmo.setMat4( 'u_tView', camera.matView() )
        shaderGizmo.setMat4( 'u_tProj', camera.matProj() )
        shaderGizmo.setVec3( 'u_color', [ 1.0, 1.0, 1.0 ] )

        gizmo.render()

        shaderGizmo.unbind()
        ##########################################################################

        #### tr.DebugDrawer.DrawNormals( mesh, [ 0.0, 0.0, 1.0 ] )
        tr.DebugDrawer.Render( camera )

        app.end()
